dht20/dht20.go

```diff
@@ -0,0 +1,103 @@
+// Package dht20 implements a driver for the DHT20 temperature and humidity sensor.
+//
+// Datasheet: https://cdn-shop.adafruit.com/product-files/5183/5193_DHT20.pdf
+
+package dht20
+
+import (
+	"time"
+
+	"tinygo.org/x/drivers"
+)
+
+// Device wraps an I2C connection to a DHT20 device.
+type Device struct {
+	bus            drivers.I2C
+	Address        uint16
+	data           [8]uint8
+	temperature    float32
+	humidity       float32
+	prevAccessTime time.Time
+}
+
+// New creates a new DHT20 connection. The I2C bus must already be
+// configured.
+//
+// This function only creates the Device object, it does not touch the device.
+func New(bus drivers.I2C) Device {
+	return Device{
+		bus:     bus,
+		Address: defaultAddress, // Using the address defined in registers.go
+	}
+}
+
+// Configure sets up the device for communication and initializes the registers if needed.
+func (d *Device) Configure() {
+	// Get the status word
+	d.data[0] = 0x71
+	d.bus.Tx(d.Address, d.data[:1], d.data[:1])
+	if d.data[0] != 0x18 {
+		// Initialize registers
+		d.initRegisters()
+	}
+	// Set the previous access time to the current time
+	d.prevAccessTime = time.Now()
+}
+
+// initRegisters initializes the registers 0x1B, 0x1C, and 0x1E to 0x00.
+func (d *Device) initRegisters() {
+	// Initialize register 0x1B
+	d.data[0] = 0x1B
+	d.data[1] = 0x00
+	d.bus.Tx(d.Address, d.data[:2], nil)
+
+	// Initialize register 0x1C
+	d.data[0] = 0x1C
+	d.data[1] = 0x00
+	d.bus.Tx(d.Address, d.data[:2], nil)
+
+	// Initialize register 0x1E
+	d.data[0] = 0x1E
+	d.data[1] = 0x00
+	d.bus.Tx(d.Address, d.data[:2], nil)
+}
+
+// Update reads data from the sensor and updates the temperature and humidity values.
+func (d *Device) Update(which drivers.Measurement) error {
+	// Check if 80ms have passed since the last access
+	if d.prevAccessTime.Add(80 * time.Millisecond).Before(time.Now()) {
+		// Check the status word Bit[7]
+		d.data[0] = 0x71
+		d.bus.Tx(d.Address, d.data[:1], d.data[:1])
+		if (d.data[0] & 0x80) == 0 {
+			// Read 7 bytes of data from the sensor
+			d.bus.Tx(d.Address, nil, d.data[:7])
+			rawHumidity := uint32(d.data[1])<<12 | uint32(d.data[2])<<4 | uint32(d.data[3])>>4
+			rawTemperature := uint32(d.data[3]&0x0F)<<16 | uint32(d.data[4])<<8 | uint32(d.data[5])
+
+			// Convert raw values to human-readable values
+			d.humidity = float32(rawHumidity) / 1048576.0 * 100
+			d.temperature = float32(rawTemperature)/1048576.0*200 - 50
+
+			// Trigger the next measurement
+			d.data[0] = 0xAC
```

Comment:
If the measurement is only triggered here that means the data that was read before is from previous measurement. If the user cod is calling `Update` once a minute (or on even longer interval) it would be always getting stale data. Would it be better to just wait 80ms until update is finished and then read the values? Other option would be to change the API, for example to have separate methods for triggering update and reading the data.

Reply:
The use of Update() is aligned with the interface in #345. I feel that waiting for 80ms within Update() is problematic. As you mentioned, the current code results in obtaining stale data. We would like to discuss how to proceed in a separate PR or other discussion.

For now, I have added a note in the comments.

Reply:
The code is marked as "outdated" by github. Is the issue you all were discussing still present?

My only gripe with the code is that Update does not actually update the sensor and returns no indication of such. We could have a sentinel error for this case or choose to block until update finishes. Both are not without their pros and cons.

Reply:
That said, now come to think of it the sentinel error might be the most idiomatic way to solve this without requiring users to start using goroutines to manage several sensors which have a cooldown period.

Reply:
@soypat the issue I noted is that `Update` first reads the data from the sensor and stores it into `Device` object, and then triggers next measurement, so when user code reads the temperature it is the value that was not measured after last `Update` call, but after second to last `Update` call and hance stail. Maybe `Update` should just trigger the measurement and `Temperature` and `Humidity` methods should check if enough time has passed (or read the status byte) and if so, do the reading from the sensor and store the value in `Device`, all subsequent calls could just return stored values. Those methods should also return error in this case. I also agree that Update should return sentinel error if it is called before enough time has passed. 
Another thing I just noticed is that `which` argument of `Update` is ignored and both Temperature and Humidity are always updated. What if `Update` is called with `drivers.Humidity`? Should the Temperature also be updated?

Reply:
ah, that sounds like it could cause several headaches. I'd document the Sensor interface such that the function returns a nil error upon starting and ending a new sensor measurment. 

* Temperature and Humidity need not check anything, those methods always return stale data (to some extent). If Staleness functionality should be desired on the Humidity/Temperature methods it probably be implemented in a type that contains a Sensor type and a callback to the Temperature/Humidity/Pressure/Distance method, that way it is readily reusable among different sensors.

* `which` should not be ignored. The simplest implementation should perform I/O only if one or more of the corresponding measurements are set. It is OK for it to update both Temperature and Humidity if only Humidity or Temperature is passed. But if neither Temp nor Humidity are set Update should return immediately with a nil error.

Reply:
I have added the check for which.